http_server_proxy.go

```diff
@@ -0,0 +1,186 @@
+	// HTTPServerProxyStoreModePassthrough mode
+	HTTPServerProxyStoreModePassthrough HTTPServerProxyStoreMode = "passthru"
+)
+
+// HTTPServerProxyStoreRequestData definition
+type HTTPServerProxyStoreRequestData struct {
+	Method  string      `json:"method"`
+	Path    string      `json:"path"`
+	Headers http.Header `json:"header"`
+	Query   url.Values  `json:"query"`
+	Body    interface{} `json:"body"`
+}
+
+// HTTPServerProxyStoreResponseData definition
+type HTTPServerProxyStoreResponseData struct {
+	StatusCode int         `json:"statuscode"`
+	Headers    http.Header `json:"header"`
+	Body       interface{} `json:"body"`
+}
+
+// HTTPServerProxyStoreDataEntry definition
+type HTTPServerProxyStoreDataEntry struct {
+	Offset   int                              `json:"offset"`
+	Request  HTTPServerProxyStoreRequestData  `json:"request"`
+	Response HTTPServerProxyStoreResponseData `json:"response"`
+}
+
+// HTTPServerProxyStore definition
+type HTTPServerProxyStore struct {
+	Name string                   `json:"name"`
+	Mode HTTPServerProxyStoreMode `json:"mode"`
+	Data []HTTPServerProxyStoreDataEntry
+}
+
+// HTTPServerProxyStoreResponse definition
+type HTTPServerProxyStoreResponse struct {
+	Mode       HTTPServerProxyStoreMode        `json:"mode"`
+	Offset     int                             `json:"offset"`
+	NextOffset int                             `json:"next_offset"`
+	Limit      int                             `json:"limit"`
+	Count      int                             `json:"count"`
+	Store      []HTTPServerProxyStoreDataEntry `json:"store"`
+}
+
+// HTTPServerProxy definition
+type HTTPServerProxy map[string]HTTPServerProxyStore
+
+// NewProxy allocates a new http proxy from its configuration
+func NewProxy(cfg HTTPServerProxy) *HTTPServerProxy {
+	proxy := new(HTTPServerProxy)
+	*proxy = cfg
+	return proxy
+}
+
+// Listen for the proxy store/retrieve
+func (proxy *HTTPServerProxy) Listen(mux *http.ServeMux, prefix string) {
+	for p, s := range *proxy {
+		store := HTTPServerProxyStore{Name: p, Mode: s.Mode, Data: []HTTPServerProxyStoreDataEntry{}}
+		mux.HandleFunc(fmt.Sprintf("%s/%s", prefix, p), store.write)
+		mux.HandleFunc(fmt.Sprintf("%sstore/%s", prefix, p), store.read)
+	}
+}
+
+// write stores incoming request data
+func (store *HTTPServerProxyStore) write(w http.ResponseWriter, r *http.Request) {
+
+	reqData := HTTPServerProxyStoreRequestData{
+		Method:  r.Method,
+		Path:    r.URL.RequestURI(),
+		Headers: r.Header,
+		Query:   r.URL.Query(),
+	}
+
+	w.Header().Add("Content-Type", "application/json")
+
+	if r.Body != nil {
+		bin, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			w.WriteHeader(500)
+			errStr := fmt.Sprintf("{\"error\":\"Could not read request body: %s\"}", err)
+			w.Write([]byte(errStr))
+			return
+		}
+		if len(bin) > 0 {
+			err = json.Unmarshal(bin, &reqData.Body)
+			if err != nil {
+				w.WriteHeader(500)
+				errStr := fmt.Sprintf("{\"error\":\"Could not unmarshal request body: %s\"}", err)
```

Comment:
Never use http status codes directly. Use `http.InternalServerError` if needed. But I don't like the whole block & JSON parsing here.